runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/PuzzleState.java

```diff
@@ -179,4 +180,45 @@ public int getHeuristicValue(Heuristic heuristic)
 
 		return h;
 	}
+
+	public PuzzleState swap(Point p1, Point p2)
+	{
+		int[] pieces = this.pieces.clone();
```

Comment:
This cloning should be after the check. Also use this for cloning instead of cloning the array itself.
```java
PuzzleState newState = new PuzzleState(this);

newState.pieces[p1.getY() * DIMENSION + p1.getX()] = val2;
newState.pieces[p2.getY() * DIMENSION + p2.getX()] = val1;

return newState;
```

Reply:
Had to set the empty piece also.